dev/set_matrix.py

```diff
@@ -0,0 +1,489 @@
+"""
+A script to set a matrix for the cross version tests for MLflow Models / autologging integrations.
+
+# How to run:
+
+```
+# ===== Include all items =====
+
+python dev/set_matrix.py
+
+# ===== Include only `ml-package-versions.yml` updates =====
+
+REF_VERSIONS_YAML="https://raw.githubusercontent.com/mlflow/mlflow/master/ml-package-versions.yml"
+python dev/set_matrix.py --ref-versions-yaml $REF_VERSIONS_YAML
+
+# ===== Include only flavor file updates =====
+
+CHANGED_FILES="
+mlflow/keras.py
+mlflow/tensorlfow/__init__.py
+"
+python dev/set_matrix.py --changed-files $CHANGED_FILES
+
+# ===== Include both `ml-package-versions.yml` & flavor file updates =====
+
+python dev/set_matrix.py --ref-versions-yaml $REF_VERSIONS_YAML --changed-files $CHANGED_FILES
+```
+
+# How to run doctests:
+
+```
+pytest dev/set_matrix.py --doctest-modules --verbose
+```
+"""
+
+import argparse
+from distutils.version import LooseVersion
+import functools
+import json
+import operator
+import os
+import re
+import shutil
+import urllib.request
+
+import yaml
+
+VERSIONS_YAML_PATH = "ml-package-versions.yml"
+DEV_VERSION = "dev"
+
+
+def read_yaml(location, if_error=None):
+    """
+    Reads a YAML file.
+
+    Examples
+    --------
+    >>> read_yaml("https://raw.githubusercontent.com/mlflow/mlflow/master/.circleci/config.yml")
+    {...}
+    >>> read_yaml(".circleci/config.yml")
+    {...}
+    >>> read_yaml("non_existent.yml", if_error={})
+    Failed to read ...
+    {}
+    """
+    try:
+        if re.search("^https?://", location):
+            with urllib.request.urlopen(location) as f:
+                return yaml.load(f, Loader=yaml.SafeLoader)
+        else:
+            with open(location) as f:
+                return yaml.load(f, Loader=yaml.SafeLoader)
+    except Exception as e:
+        if if_error is not None:
+            print("Failed to read '{}' due to: `{}`".format(location, e))
+            return if_error
+
+        raise
+
+
+@functools.lru_cache(maxsize=None)
```

Comment:
What's the benefit of the lru cache here?

Reply:
It's just for a speedup. If we didn't cache, we'd query `https://pypi.python.org/pypi/{package_name}/json` four times (("models" + "autologging") x (`config` + `config_ref`)).

Reply:
I'll remove this because the benefit is too subtle.